docs/reference/eql/functions.asciidoc

```diff
@@ -0,0 +1,60 @@
+[[eql-function-ref]]
+== EQL function reference
+++++
+<titleabbrev>Function reference</titleabbrev>
+++++
+
+experimental::[]
+
+{es} supports the following EQL functions:
+
+* <<eql-substring-fn>>
+
+[discrete]
+[[eql-substring-fn]]
+=== `substring`
+Extracts a substring from a source string at provided start and end positions.
+
+If no end position is provided, the function extracts only the character in the
+start position.
+
+[%collapsible]
+====
+*Example*
+[source,eql]
+----
+substring("quick brown fox", 0, 5)      // returns "quick"
+substring("quick brown fox", 6, 11)     // returns "brown"
+substring("quick brown fox", 6)         // returns "b"
+substring("quick brown fox", -3, -1)    // returns "fo"
+substring("quick brown fox", -3)        // returns "f"
+----
```

Comment:
@costin 
These examples are a modified version of those from the [original EQL substring docs][0].

However, they're not all working as expected in ES.

I included my testing approach below. Let me know if this behavior is expected. If so, I can update the docs.

### Index setup

Map `process.name` as a `keyword` field so it doesn't get analyzed.
```
PUT my_index
{
  "mappings": {
    "properties": {
      "process": {
        "properties": {
          "name": {
            "type": "keyword"
          }
        }
      }
    }
  }
}
```
Index a doc containing `quick brown fox` in the `process.name` field.
```
PUT my_index/_doc/1
{
  "@timestamp": "2020-12-06T11:04:05.000Z",
  "event": {
    "category": "process"
  },
  "process": {
    "name": "quick brown fox"
  }
}
```

### Starting position of `0` returns results

A starting position of `0` returns results as expected.
```
GET my_index/_eql/search
{
  "query": """
    process where substring(process.name, 0, 5) == "quick"
  """
}

GET my_index/_eql/search
{
  "query": """
    process where substring(process.name, 0, 7) == "quick b"
  """
}
```

Response:

```
{
  "took" : 2,
  "timed_out" : false,
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "events" : [
      {
        "_index" : "my_index",
        "_id" : "1",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2020-12-06T11:04:05.000Z",
          "event" : {
            "category" : "process"
          },
          "process" : {
            "name" : "quick brown fox"
          }
        },
        "sort" : [
          1607252645000
        ]
      }
    ]
  }
}
```

### Non-zero, positive offsets return no results

```
GET my_index/_eql/search
{
  "query": """
    process where substring(process.name, 6, 11) == "brown"
  """
}

GET my_index/_eql/search
{
  "query": """
    process where substring(process.name, 6) == "b"
  """
}
```

Response:
```
{
  "took" : 1,
  "timed_out" : false,
  "hits" : {
    "total" : {
      "value" : 0,
      "relation" : "eq"
    },
    "events" : [ ]
  }
}
```

### A negative starting position with no end position returns no results
```
GET my_index/_eql/search
{
  "query": """
    process where substring(process.name, -3) == "f"
  """
}
```

Response:
```
{
  "took" : 1,
  "timed_out" : false,
  "hits" : {
    "total" : {
      "value" : 0,
      "relation" : "eq"
    },
    "events" : [ ]
  }
}
```

### A negative end position returns an error
```
GET my_index/_eql/search
{
  "query": """
    process where substring(process.name, -3, -1) == "fo"
  """
}
```

Error:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "sql_illegal_argument_exception",
        "reason" : "A positive number is required for [length]; received [-1]"
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "my_index",
        "node" : "UYH0jOvXSN-ot5Vu6ml-_w",
        "reason" : {
          "type" : "script_exception",
          "reason" : "runtime error",
          "script_stack" : [
            "org.elasticsearch.xpack.sql.expression.function.scalar.string.SubstringFunctionProcessor.doProcess(SubstringFunctionProcessor.java:63)",
            "org.elasticsearch.xpack.sql.expression.function.scalar.whitelist.InternalSqlScriptUtils.substring(InternalSqlScriptUtils.java:418)",
            "InternalSqlScriptUtils.nullSafeFilter(InternalSqlScriptUtils.eq(InternalSqlScriptUtils.substring(InternalSqlScriptUtils.docValue(doc,params.v0),params.v1,params.v2),params.v3))",
            "                                                                                                                                                                ^---- HERE"
          ],
          "script" : "InternalSqlScriptUtils.nullSafeFilter(InternalSqlScriptUtils.eq(InternalSqlScriptUtils.substring(InternalSqlScriptUtils.docValue(doc,params.v0),params.v1,params.v2),params.v3))",
          "lang" : "painless",
          "position" : {
            "offset" : 160,
            "start" : 0,
            "end" : 176
          },
          "caused_by" : {
            "type" : "sql_illegal_argument_exception",
            "reason" : "A positive number is required for [length]; received [-1]"
          }
        }
      }
    ],
    "caused_by" : {
      "type" : "sql_illegal_argument_exception",
      "reason" : "A positive number is required for [length]; received [-1]"
    }
  },
  "status" : 400
}
```

[0]: https://eql.readthedocs.io/en/latest/query-guide/functions.html#substring